### PR TITLE
[client] Use a dedicated connection for nftables ipset operations

### PR DESCRIPTION
**Pull request by Laotree**
## Describe your changes

Fixes the issue reported in https://github.com/netbirdio/netbird/discussions/7446 where the client fails to install inbound peer ACL rules with:

```
flush: conn.Receive: netlink receive: no such file or directory
```

Since #6322 (`778b3b326`) the nftables backend routes both **named ipset** operations and **rule** installation through a single `family.conn`. Large set-element batches and rule flushes sharing one netlink stream can overload a batch and desync the kernel ack stream (google/nftables#170). The failure then surfaces as a spurious `no such file or directory` when the rule that references the just-created source-IP set is flushed, so the inbound peer ACL never lands.

Before #6322 the code deliberately used a separate connection for set operations, with a comment explaining that using one connection for both "overloads netlink".

This change restores a dedicated `sConn` on `family` for named ipset create/update/delete, keeping set-element operations off the rule netlink stream. #6322's atomic install of the peer filter rule and its paired mangle rule is preserved: both still commit on `conn` in a single flush. Anonymous port sets remain on `conn` because they bind atomically to their rule.

## Issue ticket number and link

No linked GitHub issue; the fix originates from https://github.com/netbirdio/netbird/discussions/7446.

### Checklist
- [x] Is it a bug fix
- [ ] Is a typo/documentation fix
- [ ] Is a feature enhancement
- [ ] It is a refactor
- [x] Created tests that fail without the change (if possible)
- [ ] I ran and tested this change locally — I did not rely on CI to find out whether it works
- [x] This PR has a single purpose (not a fix + refactor + feature in one)
- [x] This change is a trivial fix, **OR** it links an issue the NetBird team agreed on beforehand. Changes to the public API, gRPC protocols, functionality behavior, CLI / service flags, or new features always need that agreement first. See [CONTRIBUTING.md](https://github.com/netbirdio/netbird/blob/main/CONTRIBUTING.md#ticket-first-pr-second).

> By submitting this pull request, you confirm that you have read and agree to the terms of the [Contributor License Agreement](https://github.com/netbirdio/netbird/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT.md).

## Documentation
Select exactly one:

- [ ] I added/updated documentation for this change
- [x] Documentation is **not needed** for this change (explain why)

This is an internal firewall-backend refactor that does not change user-facing behavior, so no documentation is required.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved nftables firewall reliability when creating, updating, and removing rules and named IP sets.
  * Improved consistency between firewall rule changes and IP set updates, including safer recovery after failed operations.
  * Improved batching, cleanup, and retry handling for pending IP set element changes.
  * Improved reliability of IPv4 and IPv6 IP set operations.
  * Improved support for multi-source peer filters backed by named source sets.
  * Improved handling of partially completed firewall updates to prevent duplicate IP set changes.

### CODE REVIEW

**Comment by pappz** [comment]
@cubic-dev-ai review this PR

**Comment by Laotree** [comment]
## Regression test — inbound ACL / DNS forwarder

**Environment:** Debian 12 · kernel 6.1.158+ · nftables 1.0.6  
(`CONFIG_NFT_FIB_IPV4` and `CONFIG_NFT_FIB_IPV6` are unset)

**Repro (`8b129c7e1`, set-connection split only):** fails at startup
- `apply peer ACLs: flush: netlink receive: no such file or directory`
- `total rules count: 0`
- DNS forwarder fails
- `netbird-acl-input-rules` holds only `established,related`

**Verify (`bd5e36fdf`, includes the fib-mangle split):** clean
- ACL applies: `total rules count: 1`
- set-referencing rule lands (`ip saddr @nb-… accept` plus `udp/tcp 22054`)
- DNS forwarder starts and answers (port 53/22054)
- no startup ENOENT on the **input** ACL flush

**Residual:** `flush prerouting mangle rule: no such file or directory` still appears a few times at startup. That is the prerouting pair using `nft_fib`, which this kernel does not provide. It is flushed in a separate batch so it no longer rolls back inbound ACLs or DNS. NAT mangle marks (no fib) still install. Redirect-mark pairing for DNAT-to-host traffic is still missing on this kernel.

**Result:** `bd5e36fdf` fixes inbound ACL and DNS forwarder on this host.

File: client/firewall/nftables/family_conn_linux_test.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+		name string
+		r    *family
+	}{
+		{
+			name: "IPv4",
+			r:    &family{conn: &nftables.Conn{}, sConn: &nftables.Conn{}, af: afIPv4},
+		},
+		{
+			name: "IPv6",
+			r:    &family{conn: &nftables.Conn{}, sConn: &nftables.Conn{}, af: afIPv6},
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/family_conn_linux_test.go]: Not changing. The test already goes through newFamily (conn vs sConn plus pendingSetElements). A struct-literal assertion would be a tautology; this one is not.

File: client/firewall/nftables/family_conn_linux_test.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			require.NotNil(t, tc.r.conn, "rule connection must not be nil")
+			require.NotNil(t, tc.r.sConn, "set connection must not be nil")
+			assert.NotSame(
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/family_conn_linux_test.go:49]: Not changing. The test already calls newFamily rather than constructing family{conn, sConn} by hand. Asserting those two pointers differ is the production invariant, not a tautology.

File: client/firewall/nftables/filter_linux.go [hunk 3]
```diff
@@ -99,10 +82,22 @@ func (r *family) AddFilterRule(
...
+	// peer rules hit this path with no named set, so the set-ID fix
+	// cannot save them.
 	if err := r.conn.Flush(); err != nil {
+		r.discardPendingSetElements()
 		r.dropNetworkMatch(exprs)
 		return nil, fmt.Errorf(flushError, err)
 	}
+	if err := r.commitPendingSetElements(); err != nil {
+		log.Errorf("add remaining ipset elements after rule flush: %v", err)
+	}
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/filter_linux.go:105]: Not changing the success return. After Flush the filter rule and the first 1500 prefixes are already in the kernel. Returning an error without tracking that live rule orphans it; a retry then hits EEXIST on createIpSet. Logging keeps the ACL. Overflow retries now keep remaining pending entries (6065128bc).

File: client/firewall/nftables/filter_linux.go [hunk 4]
```diff
@@ -99,10 +82,22 @@ func (r *family) AddFilterRule(
...
+
+	mangleRule := r.flushPreroutingPair(srcExprs, proto, sPort, dPort, userData, isRoute)
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/filter_linux.go:107]: Not changing. Swallowing the mangle flush is the production fix: kernels without nft_fib return ENOENT and a shared batch used to roll back the ACL (empty netbird-acl-input-rules). We cannot tell fib-missing from a 'transient' ENOENT. Peer ACL and DNS forward stay; the residual mangle log is expected on those hosts.

File: client/firewall/nftables/ipset_linux.go [hunk 5]
```diff
@@ -49,40 +55,49 @@ func (r *family) createIpSet(setName string, input setInput) (*nftables.Set, err
...
+// rule batch. The named set must already exist in the kernel.
+func (r *family) commitPendingSetElements() error {
+	pending := r.pendingSetElements
+	r.pendingSetElements = make(map[string]pendingSetUpdate)
+	maxElements := maxPrefixesSet * 2
+	for setName, p := range pending {
+		if err := r.addElementBatches(p.set, p.elements, maxElements); err != nil {
+			return fmt.Errorf("add remaining elements to set %s: %w", setName, err)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/ipset_linux.go]: Fixed in 6065128bc. commitPendingSetElements now keeps remaining overflow entries until their batches succeed.

File: client/firewall/nftables/filter_linux.go [hunk 6]
```diff
@@ -99,10 +82,22 @@ func (r *family) AddFilterRule(
...
+	// expression is missing the kernel returns ENOENT and a shared batch
+	// would roll back the ACL as well. DNS forward and single-source
+	// peer rules hit this path with no named set, so the set-ID fix
+	// cannot save them.
 	if err := r.conn.Flush(); err != nil {
+		r.discardPendingSetElements()
 		r.dropNetworkMatch(exprs)
 		return nil, fmt.Errorf(flushError, err)
 	}
+	if err := r.commitPendingSetElements(); err != nil {
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/filter_linux.go:103]: Not propagating as a failed AddFilterRule. The rule and the first chunk are already committed; treating overflow failure as install failure without rollback tracking leaves a live untracked ACL. Log and keep the rule. Remaining overflow is now retained for a later commit.

File: client/firewall/nftables/filter_linux.go [hunk 7]
```diff
@@ -117,6 +112,36 @@ func (r *family) AddFilterRule(
...
+	mangleExprs, err := r.buildPeerFilterExprs(srcExprs, proto, sPort, dPort)
+	if err != nil {
+		log.Errorf("build mangle rule: %v", err)
+		return nil
+	}
+	mangleRule := r.queuePreroutingRule(mangleExprs, userData)
+	if mangleRule == nil {
+		return nil
+	}
+	if err := r.conn.Flush(); err != nil {
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/filter_linux.go:156]: Not changing. Returning error here reintroduces the Debian 12 / no-nft_fib failure: empty input ACL and DNS forward never starts. The mangle pair is best-effort when nft_fib is unavailable; the filter ACL is the must-keep path.

File: client/firewall/nftables/ipset_linux.go [hunk 8]
```diff
@@ -49,40 +55,49 @@ func (r *family) createIpSet(setName string, input setInput) (*nftables.Set, err
...
 
-// addRemainingElements adds element batches beyond the initial one in
-// maxElements-sized chunks, flushing each. Called after the set has been
-// created with its first batch.
-func (r *family) addRemainingElements(nfset *nftables.Set, elements []nftables.SetElement, maxElements int) error {
+// commitPendingSetElements writes overflow chunks that did not fit in the
+// rule batch. The named set must already exist in the kernel.
+func (r *family) commitPendingSetElements() error {
+	pending := r.pendingSetElements
+	r.pendingSetElements = make(map[string]pendingSetUpdate)
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/ipset_linux.go]: Fixed in 6065128bc. Failed sets stay in pendingSetElements instead of being dropped before the write.

File: client/firewall/nftables/ipset_linux.go [hunk 9]
```diff
@@ -49,40 +55,49 @@ func (r *family) createIpSet(setName string, input setInput) (*nftables.Set, err
...
-	log.Debugf("Created new ipset: %s with %d initial prefixes (total prefixes %d)", setName, len(initialElements)/2, len(prefixes))
-
-	// The set is committed now. If a later batch fails, destroy it: the
-	// refcounter records nothing on a create-callback error, so it would
-	// otherwise leak, and a partial source set fails-open for deny rules.
-	if err := r.addRemainingElements(nfset, elements, maxElements); err != nil {
-		if derr := r.deleteIpSet(setName, nfset); derr != nil {
-			log.Warnf("rollback ipset %s after add failure: %v", setName, derr)
+	if nElements > maxElements {
+		r.pendingSetElements[setName] = pendingSetUpdate{
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/ipset_linux.go:61]: Fixed in 6065128bc. AddNatRule now commits pending overflow after the NAT rule flush (and discards pending if that flush fails).

File: client/firewall/nftables/family_linux.go [hunk 10]
```diff
@@ -69,16 +69,32 @@ type setInput struct {
...
-	filterTable *nftables.Table
-	chains      map[string]*nftables.Chain
+	conn *nftables.Conn
+	// sConn is used for element updates and deletes of named sets that
+	// already exist in the kernel. Creating a named set that a rule will
+	// look up must be queued on conn and flushed with that rule: the
+	// lookup's SetID is valid only in the creating transaction, and a
+	// later rule batch that names a set created on another (or prior)
+	// transaction is rejected with ENOENT on some kernels.
+	sConn *nftables.Conn
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/family_linux.go:90]: Fixed in 6065128bc. AddFilterRule build errors now call rollbackQueuedNetwork: discard pending overflow, Flush conn so the queued NEWSET lands, then dropNetworkMatch so sConn can delete it.

File: client/firewall/nftables/family_conn_linux_test.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
+package nftables
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/family_conn_linux_test.go:3]: Fixed in 6065128bc. family_conn_linux_test.go is gated with //go:build !android && privileged like the other newFamily tests.

File: client/firewall/nftables/filter_linux.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/filter_linux.go:107]: Not changing. google/nftables always sends SetID; the kernel looks up by name first, and ID is same-batch only. After the first flush the named set exists, so the second mangle flush should resolve by name where nft_fib exists. The host failure we reproduced was fib ENOENT, not a stale SetID.

File: client/firewall/nftables/routing_linux.go [hunk 12]
```diff
@@ -65,9 +65,14 @@ func (r *family) AddNatRule(pair firewall.RouterPair) error {
 	}
 
 	if err := r.conn.Flush(); err != nil {
+		r.discardPendingSetElements()
 		r.rollbackRules(pair)
 		return fmt.Errorf("insert rules for %s: %w", pair.Destination, err)
 	}
+	if err := r.commitPendingSetElements(); err != nil {
+		r.rollbackRules(pair)
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/routing_linux.go]: Fixed in 1e37a0f4a by keeping the live NAT rules instead of deleting tracking. After Flush the kernel already has the rules; rollbackRules() only dropped r.rules, which leaked NAT the route manager never recorded. Same as AddFilterRule: log overflow, keep the tracked rules, retry remaining elements later. We are not deleting the kernel rules on overflow: the first 1500 prefixes are already in the set, and a failed delete would leave the same leak.

File: client/firewall/nftables/ipset_linux.go [hunk 13]
```diff
@@ -49,40 +57,57 @@ func (r *family) createIpSet(setName string, input setInput) (*nftables.Set, err
...
+	if len(r.pendingSetElements) == 0 {
+		return nil
+	}
+
+	maxElements := maxPrefixesSet * 2
+	remaining := make(map[string]pendingSetUpdate)
+	var merr *multierror.Error
+	for setName, p := range r.pendingSetElements {
+		if err := r.addElementBatches(p.set, p.elements, maxElements); err != nil {
+			remaining[setName] = p
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/ipset_linux.go]: Fixed in 1e37a0f4a. addElementBatches now returns the uncommitted suffix; commitPendingSetElements stores only that slice.

File: client/firewall/nftables/routing_linux.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/routing_linux.go]: Fixed in 1e37a0f4a. We keep tracking and return success after a successful Flush (log the overflow), matching AddFilterRule. Returning error after rollbackRules leaked kernel NAT because addToServerNetwork never recorded the route. Remaining overflow is retried on a later commit.

File: client/firewall/nftables/filter_linux.go [hunk 14]
```diff
@@ -53,7 +53,7 @@ func (r *family) AddFilterRule(
 		exprs, err = r.buildPeerFilterExprs(srcExprs, proto, sPort, dPort)
 	}
 	if err != nil {
-		r.dropNetworkMatch(srcExprs)
+		r.rollbackQueuedNetwork(srcExprs)
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/filter_linux.go]: Fixed in 1e37a0f4a. buildRouteFilterExprs no longer decrements the destination set on proto/port errors; AddFilterRule rolls back the returned exprs (source and destination) after flushing the queued NEWSET.

File: client/firewall/nftables/filter_linux.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/filter_linux.go:105]: Fixed in 1e37a0f4a. The existing-rule path now calls commitPendingSetElements so an unchanged ACL can still drain retained overflow.

**Review by pappz — CHANGES_REQUESTED** [review]
A deny rule can have more than 1500 source prefixes. The rule goes in first. The extra prefixes go in after. If that second step fails, the code only logs it. It still reports success. The rule stays live, but the missing sources are not blocked.

Main did this differently. There the same failure deleted the set and returned an error.

File: client/firewall/nftables/family_linux.go [hunk 15]
```diff
@@ -69,16 +69,32 @@ type setInput struct {
...
-	workTable   *nftables.Table
-	filterTable *nftables.Table
-	chains      map[string]*nftables.Chain
+	conn *nftables.Conn
+	// sConn is used for element updates and deletes of named sets that
+	// already exist in the kernel. Creating a named set that a rule will
+	// look up must be queued on conn and flushed with that rule: the
+	// lookup's SetID is valid only in the creating transaction, and a
+	// later rule batch that names a set created on another (or prior)
+	// transaction is rejected with ENOENT on some kernels.
```
Comment (pappz) [inline client/firewall/nftables/family_linux.go:93]: Could you share the actual netlink evidence for the ENOENT? The reason by  AI suggestion: nft_set_lookup_global() resolves by name first and only reads the set ID if the name lookup fails, so a stale SetID should be inert. If that's right, the deferred createIpSet flush isn't needed and sConn alone covers the desync.

**Comment by Laotree** [comment]
> A deny rule can have more than 1500 source prefixes. The rule goes in first. The extra prefixes go in after. If that second step fails, the code only logs it. It still reports success. The rule stays live, but the missing sources are not blocked.
> 
> Main did this differently. There the same failure deleted the set and returned an error.

Agreed. The current logging + success is misleading. 

The rule and first chunk land together on `conn`; the overflow goes through `sConn` afterwards. Returning an error like `main` does would leave a live, untracked rule behind, and a retry would then hit `EEXIST`, so the uncommitted suffix stays in `pendingSetElements` instead.

My thought is to keep the already-installed part, but return an error when the overflow commit fails, so the caller knows the ACL was not fully applied. Would that match your expectation?

**Comment by Laotree** [comment]
Also, should we add a test around the maxPrefixesSet boundary, e.g. 1501 prefixes, to cover the overflow/pending path?

File: client/firewall/nftables/routing_linux.go [hunk 16]
```diff
@@ -91,6 +103,48 @@ func (r *family) rollbackRules(pair firewall.RouterPair) {
 	}
 }
 
+// rollbackFlushedNat removes NAT/legacy rules that already landed in the
+// kernel after a successful conn.Flush, then drops their ipset refs. This
+// is the overflow-failure path: untracking alone would leak kernel rules
+// because the route manager never recorded a successful add.
+func (r *family) rollbackFlushedNat(pair firewall.RouterPair) {
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/routing_linux.go]: Agreed. `rollbackFlushedNat` now takes only the rule IDs that this `AddNatRule` actually queued. If `queueNatRule` / `queueLegacyRouteRule` skip a replacement and leave the previous live rule tracked, that key is not rolled back.

Fixed in 5c60b69c2.

File: client/firewall/nftables/filter_linux.go [hunk 17]
```diff
@@ -117,6 +122,79 @@ func (r *family) AddFilterRule(
...
+			if err := r.conn.DelRule(toDelete); err != nil {
+				log.Errorf("queue overflow rollback delete: %v", err)
+			} else if err := r.conn.Flush(); err != nil {
+				log.Errorf("flush overflow rollback delete: %v", err)
+			}
+		} else {
+			log.Errorf("overflow rollback: filter rule has no handle, kernel rule may leak")
+		}
+	}
+	r.dropNetworkMatch(exprs)
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/filter_linux.go]: Agreed. `rollbackFlushedFilterRule` now drops ipset refs only after `deleteFlushedRule` confirms the kernel delete. If the delete does not commit, the refs stay so a live rule cannot point at a set the counter already released.

Fixed in 5c60b69c2.

File: client/firewall/nftables/ipset_linux.go [hunk 18]
```diff
@@ -49,44 +57,135 @@ func (r *family) createIpSet(setName string, input setInput) (*nftables.Set, err
...
+// commitPendingSets writes overflow chunks for the named sets, retrying
+// immediately a few times. Entries stay in the map until their batches
+// succeed so a retry does not replay prefixes that already landed.
+func (r *family) commitPendingSets(names []string) error {
+	if len(names) == 0 {
+		return nil
+	}
+	var err error
+	for attempt := 1; attempt <= pendingSetCommitAttempts; attempt++ {
+		err = r.commitPendingSetsOnce(names)
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/ipset_linux.go:81]: Not taking this as a code change.

An ambiguous `Flush` error is inherent to nftables batching: the kernel may have applied the batch even when netlink returns an error. That was already true of the pre-retry chunk path in `createIpSet` / `addElementBatches`.

Replaying the uncommitted suffix is the same contract as before. If a replay hits a duplicate interval, overflow commit fails and we fail-closed (roll the rule back) instead of reporting a truncated set as installed. Reconciling live set contents after an ambiguous Flush is a larger change than this PR, and would not be reliable without a kernel dump of the set.

The retry loop only repeats that existing contract; it does not introduce a new silent-success path.

File: client/firewall/nftables/ipset_linux_test.go [hunk 19]
```diff
@@ -34,3 +36,141 @@ func TestConvertPrefixesToSetWildcard(t *testing.T) {
 		})
 	}
 }
+
+func pendingUpdate(name string) pendingSetUpdate {
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/ipset_linux_test.go:44]: Agreed. Added `TestAddElementBatchesReturnsUncommittedSuffix`: three batches of two elements, second flush fails, leftover is the suffix starting at the failed batch.

Fixed in 5c60b69c2.

File: client/firewall/nftables/filter_linux.go [hunk 20]
```diff
@@ -117,6 +122,100 @@ func (r *family) AddFilterRule(
...
+}
+
+// rollbackFlushedFilterRule deletes a filter rule that already landed in the
+// kernel after conn.Flush, then drops the ipset refs so a partial source set
+// is not left live. The rule is not yet tracked in r.filters. Refs are
+// dropped only after the kernel delete commits; otherwise the live rule
+// would point at a set the counter already released.
+func (r *family) rollbackFlushedFilterRule(nftRule *nftables.Rule, exprs []expr.Any) {
+	if r.deleteFlushedRule(nftRule) {
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/filter_linux.go:181]: Agreed. Discarding pending before a failed kernel delete left a truncated set that `Increment` would reuse.

Overflow pending is now dropped only when rollback actually deletes the rule. If delete fails, the rule stays tracked and the suffix stays queued; the next `AddFilterRule` commits that leftover instead of treating the first 1500 prefixes as complete.

Fixed in f52f4884a.

File: client/firewall/nftables/filter_linux.go [hunk 21]
```diff
@@ -99,10 +96,34 @@ func (r *family) AddFilterRule(
...
 		r.dropNetworkMatch(exprs)
 		return nil, fmt.Errorf(flushError, err)
 	}
+	rolledBack := false
+	if err := r.commitOverflowOrRollback(queued, func() bool {
+		rolledBack = r.rollbackFlushedFilterRule(nftRule, exprs)
+		return rolledBack
+	}); err != nil {
+		if !rolledBack {
+			r.filters[ruleID] = &Rule{
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/filter_linux.go:113]: Agreed. The placeholder omitted the prerouting pair, so a retry that only finished the set would return success without the redirect mark.

`finishIncompleteFilter` now commits leftover overflow and installs the mangle from the stored source lookup plus this call's proto/ports before returning the existing rule.

Also tightened two related holes found in the same pass: NAT `Flush` failure only untracks rules this call queued, and `DeleteFilterRule` drops that rule's pending overflow.

Fixed in 591b5cd2b.

File: client/firewall/nftables/filter_linux.go [hunk 22]
```diff
@@ -117,6 +134,128 @@ func (r *family) AddFilterRule(
...
+				return fmt.Errorf("add remaining ipset elements: %w", err)
+			}
+		}
+	}
+	if existing.mangleRule != nil || isRoute {
+		return nil
+	}
+	var srcExprs []expr.Any
+	if existing.nftRule != nil {
+		srcExprs = namedLookups(existing.nftRule.Exprs)
```
[the remark this answers is not in the document]
Comment (Laotree) [reply client/firewall/nftables/filter_linux.go]: Agreed. The stored peer rule is payload(src) + lookup. `namedLookups` kept only the lookup, so the retry mangle ran lookup against register 1 after the proto compare.

`sourceMatchExprs` now copies the original source match, including the address payload and skipping proto/port.

Fixed in 17706243a.

**Comment by Laotree** [comment]
@pappz Overflow failures are retried in the same `Add*`, then the live rule is rolled back and the error is returned — no more log-and-success with a truncated set. Pending overflow is discarded only for that operation. Please take another look.

**Review by pappz — CHANGES_REQUESTED** [review]
deleteIpSet now flushes DELSET on sConn while removeNatRule still has the referencing DELRULEs queued on conn. When the last set reference is released, the kernel still sees those rules and rejects the set deletion with EBUSY. The final rule flush succeeds, leaving the set behind with a stale refcount of 1.

Please keep DELSET in the same batch after the rule deletions, or release set references only after the rule deletes have successfully flushed. deleteLegacyRuleEntry needs the same ordering.

A regression test should add and remove a masqueraded NAT rule with a set destination, asserting successful removal and no remaining set.

File: client/firewall/nftables/family_linux.go (hunk 15 above)
Comment (Laotree) [reply client/firewall/nftables/family_linux.go:93]:
```
NLMSG_ERROR ack  @15:23:49.038435Z   (fd 37, NETLINK_NETFILTER)
  error = -ENOENT  (No such file or directory)
  echoed original message:
    nlmsg_len  = 556      nlmsg_type = 0xa06 -> NFT_MSG_NEWRULE
    nlmsg_flags= NLM_F_REQUEST|NLM_F_ACK|NLM_F_ECHO|NLM_F_CREATE|NLM_F_EXCL
    nlmsg_seq  = 1263032264
    nfgenmsg: family=2 (NFPROTO_IPV4), version=0, res_id=0
    NFTA_RULE_TABLE = "netbird"
    NFTA_RULE_CHAIN = "netbird-mangle-prerouting"
    NFTA_RULE_EXPRESSIONS (452 B) = 10 expressions:
        1 meta iifname / 2 cmp "wt0"
        3 payload / 4 cmp (ip protocol = 0x11 UDP)
        5 payload / 6 cmp (0x2656 = 9814)
        7 fib            <-- Fail here
        8 cmp (0x02) / 9 immediate mark 0x0001bd20
       10 meta
    NFTA_RULE_USERDATA = "<invalid network>-cc669b18d675e3b6"
```

**Comment by Laotree** [comment]
**@pappz** both points are addressed (`50d9840b1`):

- **Ordering**: `removeNatRule` and `deleteLegacyRuleEntry` now only queue the DELRULE and return the removed rule; `RemoveNatRule`, `RemoveAllLegacyRouteRules`, and `AddNatRule` collect those rules and call `decrementSetCounter` only after `conn.Flush()` succeeds — so the DELSET lands only once the referencing rules are gone. No more EBUSY / stale refcount 1. A failed flush is atomic (the batch never lands), so references are left intact for a retry.
- **Regression test**: `TestNftablesManager_NatRuleWithSetDestinationRemovesSet` adds and removes a masqueraded NAT rule with a named-set destination, asserting both successful removal and that the set no longer exists.

The lint failure this refactor introduced (`nilnil` on the new `(rule, error)` returns) is fixed in `94381c89b` with a sentinel error. All CI is green, including the Linux golangci-lint job; CodeRabbit and cubic report no issues. Could you re-review?

**Comment by Laotree** [comment]
**Regression test — real nftables kernel (OrbStack privileged container)**

Ran via the repo harness: `make test-privileged` spawns a `golang:1.26.7-alpine` container with `--privileged --cap-add=NET_ADMIN` on an Ubuntu Noble (arm64) host.

```console
$ PRIV_RUN=TestNftablesManager_NatRuleWithSetDestinationRemovesSet PRIV_PKGS=./client/firewall/nftables/... make test-privileged
=== RUN   TestNftablesManager_NatRuleWithSetDestinationRemovesSet
time="2026-09-21T04:46:24Z" level=info msg="Created new ipset: nb-168c8496 with 0 prefixes"
--- PASS: TestNftablesManager_NatRuleWithSetDestinationRemovesSet (0.01s)
PASS
ok  	github.com/netbirdio/netbird/client/firewall/nftables	0.015s
```

The test adds a masqueraded NAT rule with a named-set destination, asserts the set exists, removes the rule, then asserts the set is gone — confirming the DELSET now lands after the referencing rules and is no longer rejected with EBUSY.

File: client/firewall/nftables/filter_linux.go [hunk 23]
```diff
@@ -256,6 +394,7 @@ func (r *family) DeleteFilterRule(rule firewall.Rule) error {
 	}
 
 	r.dropNetworkMatch(pr.nftRule.Exprs)
+	r.discardPendingSets(r.pendingForExprs(pr.nftRule.Exprs))
```
Comment (pappz) [inline client/firewall/nftables/filter_linux.go]: Discovered by AI:

`pendingSetElements` is keyed by set name, and set names are a SHA-256 hash of the sorted source prefixes (`firewall.NewPrefixSet`). Two policies with the same source list, say the same 2000 peers on TCP/443 and on UDP/53, get different `ruleID`s but share one refcounted set. So this line does not drop "that rule's pending overflow"; it drops the overflow of every set the rule referenced, including one a different, still-incomplete rule depends on.

Concretely: rule B's overflow commit fails three times, `rollbackFlushedFilterRule` cannot delete the kernel rule, so `AddFilterRule` stores B in `r.filters` with the 500-element suffix retained for the retry. Rule A then leaves the network map and `DeleteFilterRule` runs for it, and this line removes the shared set from `pendingSetElements`. The refcount only drops 2 to 1, so the truncated set stays live. On the next `ApplyFiltering`, `finishIncompleteFilter` finds nothing pending, skips the commit, installs the mangle and returns `(existing, nil)`. B is reported installed with 1500 of 2000 prefixes. For `ActionDrop` that is a silent fail-open with no log line.

Line 383 has the same issue.

Suggestion: discard only when the refcount actually reaches zero. `dropNetworkMatch` already calls `ipsetCounter.Decrement`, which returns the resulting `Ref`, so it can collect the names whose count hit 0 and pass only those to `discardPendingSets`.

Comment (Laotree) [reply client/firewall/nftables/filter_linux.go]: Fixed in a1f87b4b6.

Pending overflow is now released only by `deleteIpSet`, when the last reference to the set drops, instead of by the per-rule delete/rollback paths. `DeleteFilterRule` no longer calls `discardPendingSets(pendingForExprs(...))` (both the stale-entry path and the normal delete), and `commitOverflowOrRollback` leaves the pending map alone, so a shared set keeps its suffix while another rule still depends on it.

Tests cover the exact scenario: `TestPendingOverflowTracksSetRefcount` (sibling delete keeps the shared suffix; the last reference releases it) and `TestCommitOverflowOrRollbackKeepsPendingOnRollback`. Both run netlink-free through the existing `testPendingFlush` seam.
